File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
// class Solution {
//     public int totalFruit(int[] fruits) {
//         HashMap<Integer,Integer> mp = new HashMap<>();
//         int l=0,max_len=0;
//         for (int r=0;r<fruits.length;r++){
//             mp.put(fruits[r],mp.getOrDefault(fruits[r],0)+1);
//             while (mp.size()>2){
//                 mp.put(fruits[l],mp.getOrDefault(fruits[l],0)-1);
//                 if (mp.get(fruits[l])==0){
//                     mp.remove(fruits[l]);
//                 }
//                 l++;

//             }
//         max_len=Math.max(max_len,r-l+1);
//         }
//         return max_len;
//     }
// }
class Solution {
    public int totalFruit(int[] fruits) {
        HashMap<Integer,Integer> mp = new HashMap<>();
        int l=0,max_len=0;
        for (int r=0;r<fruits.length;r++){
            mp.put(fruits[r],mp.getOrDefault(fruits[r],0)+1);
            if (mp.size()>2){
                mp.put(fruits[l],mp.getOrDefault(fruits[l],0)-1);
                if (mp.get(fruits[l])==0){
                    mp.remove(fruits[l]);
                }
                l++;

            }
        else max_len=Math.max(max_len,r-l+1);
        }
        return max_len;
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        dic = {}
        l=0
        count =0
        for i in range(len(fruits)):
            dic[fruits[i]] = dic.get(fruits[i],0)+1
            while (len(dic)>2):
                dic[fruits[l]] = dic.get(fruits[l],0)-1
                if(dic[fruits[l]] == 0):
                    del dic[fruits[l]]
                l+=1
            count = max(count,i-l+1)
        return count
            